The pipeline is for the analysis of multiplexed bulk RNA-seq data in C. elegans;</br>
the barcode/UMI structure follows celSeq method </br>
The input are fastq files generated by bcl2fastq (this could be done in basespase) </br>
The processing includes:</br>
(1) extract valid reads by extractValidReads.py. this scripts extract R1 with valid cell barcode (allow 1 mismatch) and attach the cell barcode&UMI to the header of corresponding R2;</br>
(2) A STAR pipeline to align valid reads to the genome. This module includes Quality Control and Genome Alignment using STAR.  </br>
.....Steps:</br>
.....1) For Quality Control, we use FastQC to create qc outputs.</br>
.....2) STAR is a program that aligns RNA-Seq reads to a genome.</br>
(3) A ESAT module to count reads for each feature  </br>

